clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -479,6 +479,12 @@ private static int parseAcks(String acksString) {
             // handling exceptions and record the errors;
             // for API exceptions return them in the future,
             // for other exceptions throw directly
+        } catch (BufferExhaustedException e) {
+            this.errors.record();
+            this.metrics.sensor("buffer-exhausted-records").record();
+            if (this.interceptors != null)
+                this.interceptors.onSendError(record, tp, e);
+            throw e;
```

Comment:
Right, it's indeed an option which is much safer. 
Still I think we better throw here by following reasons:

- It sounds semantically correct more. Callback and Future should be used for providing result of asynchronous(background) processing. However, these two TimeoutException occurrs while a KafkaProducer is still doing synchronous(foreground) processing and that(kafka producer has to do some foreground processing before it appends record to the accumulator) is the reason why a caller of `producer#send` is forced to wait until the result turns out, so the caller should receive the result of that call in a synchronous way.
- Assuming this is going to be shipped with 0.10.2.0, making a breaking change on behavior isn't preferred but allowed if we leave a correct note on "breaking changes" section.
- We can expect this breaking change is relatively less-harm with expect to most users who uses producer in sensitive situation would already using it like below. Users may see some new error logs but it still doesn't break the whole processing(maybe I'm little biased, feel free to leave object if any :D)
```java
try {
    producer.send(record, (metadata, exception) -> {
        if (exception != null) {
          // logging
        }
     });
} catch (RuntimeException/* or maybe KafkaException, TimeoutException whatever */ e) {
     // logging
}
```